GoogleDataTransportCCTSupport/GoogleDataTransportCCTSupport/ProtoSupport/proto_generator.py

```diff
@@ -0,0 +1,323 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Generates and massages protocol buffer outputs.
+
+Example usage:
+
+python GoogleDataTransportCCTSupport/GoogleDataTransportCCTSupport/ProtoSupport/build_protos.py \
+  --nanopb \
+  --protos_dir=GoogleDataTransportCCTSupport/GoogleDataTransportCCTSupport/Classes/Protos/ \
+  --pythonpath=~/Downloads/nanopb-0.3.9.2-macosx-x86/generator/ \
+  --output_dir=GoogleDataTransportCCTSupport/GoogleDataTransportCCTSupport/Classes/Protogen/
+"""
+
+from __future__ import print_function
+
+import sys
+
+import argparse
+import os
+import os.path
+import re
+import subprocess
+
+OBJC_GENERATOR='nanopb_objc_generator.py'
+
+COPYRIGHT_NOTICE = '''
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'''.lstrip()
+
+
+def main():
+  parser = argparse.ArgumentParser(
+      description='Generates proto messages.')
+  parser.add_argument(
+      '--nanopb', action='store_true',
+      help='Generates nanopb messages.')
+  parser.add_argument(
+      '--objc', action='store_true',
+      help='Generates Objective-C messages.')
+  parser.add_argument(
+      '--protos_dir',
+      help='Source directory containing .proto files.')
+  parser.add_argument(
+      '--output_dir', '-d',
+      help='Directory to write files; subdirectories will be created.')
+  parser.add_argument(
+      '--protoc', default='protoc',
+      help='Location of the protoc executable')
+  parser.add_argument(
+      '--pythonpath',
+      help='Location of the protoc python library.')
+  parser.add_argument(
+      '--include', '-I', action='append', default=[],
+      help='Adds INCLUDE to the proto path.')
+
+
+  args = parser.parse_args()
+  if args.nanopb is None and args.objc is None:
+    parser.print_help()
+    sys.exit(1)
+
+  if args.protos_dir is None:
+    root_dir = os.path.abspath(os.path.dirname(__file__))
+    args.protos_dir = os.path.join(root_dir, 'protos')
+
+  if args.output_dir is None:
+    root_dir = os.path.abspath(os.path.dirname(__file__))
+    args.output_dir = os.path.join(root_dir, 'protogen-please-supply-an-outputdir')
+
+  all_proto_files = collect_files(args.protos_dir, '.proto')
+  if args.nanopb:
+    NanopbGenerator(args, all_proto_files).run()
+
+  proto_files = remove_well_known_protos(all_proto_files)
+
+  if args.objc:
```

Comment:
You've dropped the checks for cpp here (vs the original file). That's fine; you don't need it. But if you keep it, it'll make future refactoring easier. So I guess the question is; do we want to refactor with firestore's versions of these? And potentially other components in the future? If yes, then that'd argue for keeping the cpp support in this file so that refactoring becomes trivial. If no, then that'd argue for dropping cpp support, like you've done here.

Reply:
If I'm the person doing the future refactoring, I'm fine doing it without the cpp. Conceivably, the cpp stuff will continue to evolve, whereas if I had it, it'd remain the same. That'd create harder future conflicts that our lib not having any cpp massage code, imo.

Reply:
Yeah; I was thinking kinda a near-term refactoring before the sources have a change to diverge. :)  But I'm ok with the approach you've outlined. (I'm not even 100% certain that the logic won't diverge out of necessity, thereby making a refactoring impossible.)